airflow_clickhouse_plugin/hooks/clickhouse_hook.py

```diff
@@ -43,21 +44,38 @@ def get_first(self, sql: str, parameters: dict = None) -> Tuple:
         with disconnecting(self.get_conn()) as client:
             return next(client.execute_iter(sql, params=parameters))
 
-    def get_pandas_df(self, *args, **kwargs):
-        raise NotImplementedError
+    def get_pandas_df(self, sql: str) -> pd.DataFrame:
+        query_result = self.run(sql, columnar=True, with_column_types=True)
+        df = pd.DataFrame()
+        for index, field in enumerate(query_result[1]):
+            col_name = field[0]
+            try:
+                df[col_name] = query_result[0][index]
+            except IndexError:
+                df[col_name] = ()
```

Comment:
Quite hard to read. Does it have any positive performance impact over naive implementation?

```python
rows, columns_defs = self.run(sql, with_column_types=True)
columns_names = (column_name for column_name, column_type in columns_defs)
return pd.DataFrame(rows, columns=columns)
```

Reply:
Sorry for delay. Good point. I just remember some nuances with memory when using `with_column_types` / `columnar` flags. But I checked one more time. You are right.

```
r1 = c.execute(some_query)
r2 = c.execute(some_query, columnar=True)
r3 = c.execute(some_query, with_column_types=True)

print(sys.getsizeof(r1))   # 2423088
print(sys.getsizeof(r2))   # 288
print(sys.getsizeof(r3))   # 64
```

All comments was fixed. Thank you.

Reply:
`getsizeof` is not a correct way of retrieving data size. `r1` is an array of tuples and `getsizeof` seems to be able to count its total size, including all the nested data. And somewhy is is not able to calculate the size of the data for `r2` and `r3` producing only the size allocated to the data pointers. Or I didn't get the purpose of your example.

Also this does not seem to be trustworthy that the data for the same query differs 10k times in size :)

Reply:
> `getsizeof` is not a correct way of retrieving data size. `r1` is an array of tuples and `getsizeof` seems to be able to count its total size, including all the nested data. And somewhy is is not able to calculate the size of the data for `r2` and `r3` producing only the size allocated to the data pointers. Or I didn't get the purpose of your example.
> 
> Also this does not seem to be trustworthy that the data for the same query differs 10k times in size :)

Ooh!!! -( 0_o )- I apologize. It's just a hurry. Anyway I checked again and there really is a slight difference.

```
r1 = c.execute(query, with_column_types=True)
r2 = c.execute(query, columnar=True, with_column_types=True)
# amount records 559 366
print(humanfriendly.format_size(asizeof.asizeof(r1)))   # 475.01 MB
print(humanfriendly.format_size(asizeof.asizeof(r2)))   # 443.4 MB
```

What do you think?
Thank you.

Reply:
That makes sense, thank you!